torch/nn/parallel/distributed.py

```diff
@@ -130,18 +130,23 @@ def backward(ctx, *grad_outputs):
         if static_graph_training and ctx.state_dict['num_iterations'] == 1:
             Variable._execution_engine.queue_callback(ctx.reducer._delay_all_reduce)
```

Comment:
this may need to be moved after prepare_for_backward() call, and use ctx.reduer.get_num_backward_calls_() to check whether it is first backward iteration

Reply:
Would it be simpler if we just have a boolean flag in python associated with static graph training? We always only need to enqueue it once, and only the first time, so something like `enqueued_delay_allreduce` flag should be enough and simpler here?

Reply:
Also if there are N forward calls before a backward, checking `num_backward_calls` in from C++ may result in us enqueueing this multiple times, which is not wanted

Reply:
discussed offline, we are using  ctx.reduer.get_num_backward_calls_() as suggested